main.py

```diff
@@ -157,53 +157,77 @@ def serve_codejs():
 	else: request.response.set_data("game_log('load_code: Code not found',colors.code_error)")
 	return request.response
 
+# Cache when data is prepared
+cached_data = None
+
 @app.route('/data.js',methods=['GET','POST'])
 @ndb.toplevel
 def serve_datajs():
-	domain=gdi(request); additional=""
-	geometry={}
-	for name in maps:
-		key=maps[name]["key"]
-		if maps[name].get("ignore"): continue
-		#if name=="test" and not is_sdk: key="test"
-		#logging.info(key)
-		geometry[name]=get_by_iid("map|%s"%key).info.data
-	data={
-		"version":game_version,
-		"achievements":achievements,
-		"animations":animations,
-		"monsters":monsters,
-		"sprites":sprites,
-		"maps":maps,
-		"geometry":geometry,
-		"npcs":npcs,
-		"tilesets":tilesets,
-		"imagesets":imagesets,
-		"items":items,
-		"sets":sets,
-		"craft":craft,
-		"titles":titles,
-		"tokens":tokens,
-		"dismantle":dismantle,
-		"conditions":conditions,
-		"cosmetics":cosmetics,
-		"emotions":emotions,
-		"projectiles":projectiles,
-		"classes":classes,
-		"dimensions":dimensions,
-		"images":precomputed.images,
-		"levels":levels,
-		"positions":positions,
-		"skills":skills,
-		"events":events,
-		"games":games,
-		"multipliers":multipliers,
-		#"codes":codes,
-		"docs":docs,
-		"drops":drops,
-	}
-	if request.values.get("reload"): additional="add_log('Game data reloaded','#32A3B0');\napply_backup()\n"
-	request.response.set_data("var G=%s;\n%s"%(json.dumps(data),additional))
+	global cached_data
+	
+	init_request(request)
+
+	additional=""
+
+	if request.values.get("reload"):
+		additional="add_log('Game data reloaded','#32A3B0');\napply_backup()\n"
+	else:
+		# Check the ETag
+		if request.headers.get('If-None-Match') == str(game_version):
```

Comment:
I don't think relying on the `game_version` is safe enough for the tag. We should compute a md5sum of the whole object and use that instead. (You would need to store the hash as well)

Though this could have implications on server restarts if the generated string is not stable, it's an important point that we need to check.

Reply:
Why do you believe that the `game_version` would not suffice? Most of the external clients (as far as I know) depend only on the game version to determine whether or not to fetch the file again.

Reply:
Because it is something managed by a human. Since the code can do better, it should

Reply:
No, that's just doing it properly, that's how ETags work by design

Reply:
thmsn brought up a good point in discord, that if the game mode were to change (normal -> dungeon or hardcore), the data will be different, but just the game version isn't sufficient.